helix-core/src/surround.rs

```diff
@@ -172,6 +178,17 @@ pub fn find_nth_pairs_pos(
 
     let (open, close) = get_pair(ch);
     let pos = range.cursor(text);
+    let (pos, n) = match find_type {
+        FindType::Surround(n) => (pos, n),
+        FindType::Next(n) => match search::find_nth_next(text, open, pos, n) {
```

Comment:
Using plaintext search routines isn't right here. This should be based on tree-sitter. The plaintext search gets thrown off by string literals.

In general I think reusing the surround function won't work too well here. The surround function mostly reuses other functions (match_bracket:: and search::) to find pairs and only adds logical necessary for surround. For forwards/reverse search you are looking for a pair fist and then a single (simpler) call to match_bracket/search would suffice.

I would probably be ok with not implementing the tree-sitter based forward/reverse search for now but I would definitly want to see this as a seperate function (likely directly within textobject.rs).

Reply:
Sorry not sure I'm fully following - `mi{` doesn't use treesitter as far as I can tell, so would we want `]{` to behave any differently?

Also, I decided to go with the option of finding the first char within a pair and then using `textobject_pair_surround` as this allows selection of ranges between characters where the direction is not clear from the bracket e.g. between `|` characters. `]|` works fine to jump between pairs of `|` in something like `foo|bar|baz|`, but that wouldn't work if first searching for the `|` and then calling one of the `match_brackets` functions. I'm sure I could overcome this by making a bunch of changes in `match_brackets` but I thought that would add a lot of code - do you think that option would be preferable, or is there a different approach I'm missing?